crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs

```diff
@@ -0,0 +1,3467 @@
+                    .semantic()
+                    .scopes()
+                    .ancestors(component_scope_id)
+                    .skip(1)
+                    .contains(&dependency_scope_id)
+                    || dependency.chain.len() == 1 && dependency.chain[0] == "current")
+                {
+                    continue;
+                }
+            }
+
+            ctx.diagnostic(unnecessary_outer_scope_dependency_diagnostic(
+                hook_name,
+                &dependency.name,
+                dependency.span,
+            ));
+        }
+
+        let undeclared_deps = found_dependencies.difference(&declared_dependencies).filter(|dep| {
+            if declared_dependencies.iter().any(|decl_dep| dep.contains(decl_dep)) {
+                return false;
+            }
+
+            if !is_identifier_a_dependency(&dep.name, dep.reference_id, ctx, component_scope_id) {
+                return false;
+            };
+            true
+        });
+
+        if undeclared_deps.clone().count() > 0 {
+            ctx.diagnostic(missing_dependency_diagnostic(
+                hook_name,
+                &undeclared_deps.map(Dependency::to_string).collect::<Vec<_>>(),
+                dependencies_node.span(),
+            ));
+        }
+
+        // effects are allowed to have extra dependencies
+        if !is_effect {
+            let unnecessary_deps: Vec<_> =
+                declared_dependencies.difference(&found_dependencies).collect();
+
+            // lastly, we need co compare for any unnecessary deps
+            // for example if `props.foo`, AND `props.foo.bar.baz` was declared in the deps array
+            // `props.foo.bar.baz` is unnecessary (already covered by `props.foo`)
+            declared_dependencies.iter().tuple_combinations().for_each(|(a, b)| {
+                if a.contains(b) || b.contains(a) {
+                    ctx.diagnostic(unnecessary_dependency_diagnostic(
+                        hook_name,
+                        &b.to_string(),
+                        b.span,
+                    ));
+                }
+            });
+
+            for dep in unnecessary_deps {
+                if found_dependencies.iter().any(|found_dep| found_dep.contains(dep)) {
+                    continue;
+                }
+
+                ctx.diagnostic(unnecessary_dependency_diagnostic(
+                    hook_name,
+                    &dep.to_string(),
+                    dep.span,
+                ));
+            }
+        }
+
+        for dep in declared_dependencies {
+            let Some(symbol_id) = dep.symbol_id else { continue };
+
+            if dep.chain.is_empty() && is_symbol_declaration_referentially_unique(symbol_id, ctx) {
+                ctx.diagnostic(dependency_changes_on_every_render_diagnostic(hook_name, dep.span));
+            }
+        }
+    }
+
+    fn should_run(&self, ctx: &ContextHost) -> bool {
+        ctx.source_type().is_jsx()
+    }
```

Comment:
can anyone think of a case where we would want this to run on non-jsx?

Reply:
hmm i guess this is incorrect in the case where people are using `createElement` rather than jsx

Reply:
Also lib files can be .ts files that have hooks in them with effects like https://github.com/streamich/react-use/

Reply:
ah yes good thought